performance/src/main/java/org/hibernate/validator/performance/simple/MultiLevelContainerValidation.java

```diff
@@ -0,0 +1,273 @@
+		);
+		bh.consume( violations );
+	}
+
+	@Benchmark
+	@BenchmarkMode(Mode.All)
+	@OutputTimeUnit(TimeUnit.MILLISECONDS)
+	@Fork(value = 1)
+	@Threads(50)
+	@Warmup(iterations = 10)
+	@Measurement(iterations = 50)
+	public void testMultiLevelRandomLengthLessConstraintsValidation(MultiLevelContainerState state, Blackhole bh) {
+		Set<ConstraintViolation<OtherMapContainer>> violations = state.validator.validate(
+				new OtherMapContainer( prepareOtherTestData( state.random.nextInt( MAX_MAP_ENTRIES ), state.random.nextInt( MAX_LIST_ENTRIES ) ) )
+		);
+		bh.consume( violations );
+	}
+
+	/**
+	 * Contains test classes and test data generator.
+	 */
+	static final class MultiLevelContainerDataGenerator {
+
+		private static final Random RANDOM = new Random();
+
+		private MultiLevelContainerDataGenerator() {
+		}
+
+		static Map<Optional<Cinema>, List<EmailAddress>> prepareTestData(int maxEntries, int maxListEntries) {
+			Map<Optional<Cinema>, List<EmailAddress>> map = new LinkedHashMap<>();
+			for ( int i = 0; i < maxEntries; i++ ) {
+				map.put( Optional.of( Cinema.get() ), EmailAddress.getList( maxListEntries ) );
+			}
+			return map;
+		}
+
+		static Map<Cinema, List<EmailAddress>> prepareOtherTestData(int maxEntries, int maxListEntries) {
+			Map<Cinema, List<EmailAddress>> map = new LinkedHashMap<>();
+			for ( int i = 0; i < maxEntries; i++ ) {
+				map.put( Cinema.get(), EmailAddress.getList( maxListEntries ) );
+			}
+			return map;
+		}
+
+		private static String randomString() {
+			char[] chars = new char[RANDOM.nextInt( 10 ) + 1];
+			for ( int i = 0; i < chars.length; i++ ) {
+				chars[i] = (char) RANDOM.nextInt();
+			}
+			return String.valueOf( chars );
+		}
+
+		/**
+		 * Model classes.
+		 */
+
+		static class MapContainer {
+
+			private Map<@NotNull Optional<@Valid Cinema>, List<@NotNull @Valid EmailAddress>> map;
+
+			public MapContainer(Map<@NotNull Optional<@Valid Cinema>, List<@NotNull @Valid EmailAddress>> map) {
+				this.map = map;
+			}
+
+			public Map<Optional<Cinema>, List<EmailAddress>> getMap() {
+				return map;
+			}
+		}
+
+		static class UnConstraintMapContainer {
+
+			private Map<Optional<Cinema>, List<EmailAddress>> map;
+
+			public UnConstraintMapContainer(Map<Optional<Cinema>, List<EmailAddress>> map) {
+				this.map = map;
+			}
+
+			public Map<Optional<Cinema>, List<EmailAddress>> getMap() {
+				return map;
+			}
+		}
+
+		static class OtherMapContainer {
+
+			private Map<@NotNull @Valid Cinema, List<EmailAddress>> map;
+
+			public OtherMapContainer(Map<@NotNull @Valid Cinema, List<EmailAddress>> map) {
+				this.map = map;
+			}
+
+			public Map<Cinema, List<EmailAddress>> getMap() {
+				return map;
+			}
+		}
+
+		static class Cinema {
+
+			String name;
+
+			Reference<@Valid Visitor> visitor;
+
+			Cinema(String name, Reference<Visitor> visitor) {
+				this.name = name;
+				this.visitor = visitor;
+			}
+
+			static Cinema get() {
```

Comment:
Maybe, call it `generate` instead of `get`? It's more to the point I think. Same below for `Visitor`.